Refactor energy corrections in Arkane and implement new bond additivity correction type

cgrambow opened the pull request:
### Motivation or Problem
Applying energy corrections as part of a statmech job in Arkane takes place in a really messy function and does not allow for easy extension to new bond additivity correction types.

### Description of Changes
This PR refactors energy corrections into separate modules and stores the data separately. Furthermore, it implements a new type of bond additivity correction scheme.

### Testing
Existing unit tests pass.

### Discussion
This PR provides the functionality to select either type of bond additivity correction, but it doesn't yet provide data for the new BAC type. My plan is to completely replace the current atom energies using just the calculated values at the respective levels of theory and to provide newly-derived bond additivity corrections for all levels of theory for both the existing type of BACs and the newly implemented type of BACs once we have selected a reference set of molecules and performed the quantum calculations at all the levels of theory.

This PR can be merged without the new data because it doesn't change current Arkane behavior. Alternatively, we can wait with merging until the new data is available. What do you think? In either case, the current changes are ready for review.

alongd commented on arkane/encorr/mbac.py:
```diff
@@ -0,0 +1,141 @@
+    and dictionaries of BAC parameters, return the total BAC correction
+    (should be SUBTRACTED from energy).
+
+    Note that a molecular correction term other than 0 destroys the size
```
Could you explain this point? What's a molecular correction term? Is it a general name for BACs and AECs? What are the practical implications of destroying the QC method consistency?

cgrambow replied:
Equation 3 and 4 in the Melius paper describe the molecular correction term. Basically, it is a single additional parameter that corrects for potential systematic errors in the overall electronic structure of the molecule due to the difference in spin quantum number of the molecule and summed spin quantum numbers of the atoms. The paper gives a good example of what it means to destroy size consistency: "For example, the sum of the calculated energies for hydrogen and oxygen atoms is not the same as the energy of the hydroxyl radical when the two atoms are fixed at infinity." Size consistency is a desirable property of quantum chemistry methods, but I don't actually know how much of an issue it would be in practice to have this term. So far, I haven't observed any benefit when fitting the term in addition to the other parameters so right now I'm just passing it as an argument to the function with a value of 0 and don't plan on including it in the `mbac` dictionary in `data.py`.

alongd commented on arkane/encorr/mbac.py:
```diff
@@ -0,0 +1,141 @@
+    long bonds. Use RMG for hydrogen because Open Babel can't do it for
+    mysterious reasons.
+    """
+    if len(nums) == 2 and all(n == 1 for n in nums):
```
Perhaps `if nums == [1, 1]` instead?
Why hard code for H2, or am I not reading it correctly?

alongd replied:
Will `if list(nums) == [1, 1]` work?

OK. Could you add a small comment explaining why H2 was hardcoded?

cgrambow replied:
It's mentioned in the docstring of the function. Is that not enough?

alongd replied:
Perfect! (I missed it)

alongd commented on arkane/encorr/mbac.py:
```diff
@@ -0,0 +1,141 @@
+        mol.fromXYZ(nums, coords)
+    else:
+        xyz = '{}\n\n'.format(len(nums))
+        xyz += '\n'.join('{0}  {1[0]: .10f}  {1[1]: .10f}  {1[2]: .10f}'.format(s, c) for s, c in zip(coords, nums))
```
Can you take another look here? Looks like `coords` and `nums` should be switched in the `zip`

cgrambow replied:
Oops, great catch.

alongd commented on documentation/source/users/arkane/input.rst:
```diff
@@ -28,6 +28,7 @@ Component                   Description
 ``useHinderedRotors``       ``True`` (by default) if hindered rotors are used, ``False`` if not
 ``useAtomCorrections``      ``True`` (by default) if atom corrections are used, ``False`` if not
 ``useBondCorrections``      ``True`` if bond corrections are used, ``False`` (by default) if not
+``bondCorrectionType``      ``'p'`` for Petersson-type or ``'m'`` for Melius-type bond additivity corrections
```
Could you specify what the default is?

alongd commented on documentation/source/users/arkane/input.rst:
```diff
@@ -188,7 +189,12 @@ to apply atomization energy corrections (AEC) and spin orbit corrections (SOC) f
 atom corrections can be turned off by setting ``useAtomCorrections`` to ``False``.
 
-The ``bond`` parameter is used to apply bond corrections (BC) for a given ``modelChemistry``.
+The ``bond`` parameter is used to apply bond corrections (BC) for a given ``modelChemistry`` if using Petersson-type
```
Can we change `to apply bond corrections (BC)` into `to apply bond additivity corrections (BAC)`, and then just use BAC later on?

alongd commented on documentation/source/users/arkane/input.rst:
```diff
@@ -188,7 +189,12 @@ to apply atomization energy corrections (AEC) and spin orbit corrections (SOC) f
-The ``bond`` parameter is used to apply bond corrections (BC) for a given ``modelChemistry``.
+The ``bond`` parameter is used to apply bond corrections (BC) for a given ``modelChemistry`` if using Petersson-type
+bond additivity corrections (BACs) (``bondCorrectionType = 'p'``). When using Melius-type bond additivity corrections
+(``bondCorrectionType = 'm'``), specifying bonds is not required because the molecular connectivity is automatically
```
Perhaps change `specifying bonds is not required` into `specifying bond orders is not required`

cgrambow replied:
It means that the `bonds` dictionary is not required in the input file. To highlight that, I added `` around it so that it gets parsed as plain text in the documentation.

mliu49 commented on arkane/encorr/pbac.py:
```diff
@@ -0,0 +1,73 @@
+        if symbol in params:
+            bac += count * params[symbol] * 4184.0
+        elif symbol[::-1] in params:
+            bac += count * params[symbol[::-1]] * 4184.0
```
This method will break once we start supporting elements with two letter symbols, right? Although it might be a moot point if we're deprecating this type of correction anyway.

cgrambow replied:
Didn't even think of that. I replaced it with a regex which works for all possible scenarios.

cgrambow commented on arkane/encorr/corr.py:
```diff
@@ -0,0 +1,118 @@
+
+    for symbol, count in atoms.items():
+        if symbol in atom_energies:
+            corr -= count * atom_energies[symbol] * 4.35974394e-18 * constants.Na
```
I added comments everywhere but just kept the numerical values in there for now. I'm inclined to not add unit conversions to constants.py, but am open to other suggestions.

cgrambow commented on arkane/encorr/data.py:
```diff
@@ -0,0 +1,405 @@
+
+# Atom energy corrections to reach gas-phase reference state
+# Experimental enthalpy of formation at 0 K, 1 bar for gas phase
+# See Gaussian thermo whitepaper at http://www.gaussian.com/g_whitepap/thermo.htm)
```
Turns out the link has actually changed, so I replaced it with a more up-to-date link.

cgrambow commented on arkane/encorr/data.py:
```diff
@@ -0,0 +1,405 @@
+           'Rb': 17.86, 'Ag': 66.61, 'Cd': 25.240, 'Sn': 70.50, 'I': 24.04, 'Xe': -1.481,
+           'Cs': 16.80, 'Hg': 13.19, 'Pb': 15.17}
+
+# Thermal contribution to enthalpy Hss(298 K) - Hss(0 K) reported by Gaussian thermo whitepaper
```
These terms just describe the enthalpy changes for the elements going from 0 K to 298 K. I think the Hss notation was left over from a previous version of Arkane, so I just replaced it.

cgrambow commented on arkane/encorr/data.py:
```diff
@@ -0,0 +1,405 @@
+# He, Ne, K, Ca, Ti, Cu, Zn, Ge, Br, Kr, Rb, Ag, Cd, Sn, I, Xe, Cs, Hg, and Pb are taken from CODATA
+# Codata: Cox, J. D., Wagman, D. D., and Medvedev, V. A., CODATA Key Values for Thermodynamics, Hemisphere
+# Publishing Corp., New York, 1989. (http://www.science.uwaterloo.ca/~cchieh/cact/tools/thermodata.html)
+atom_hf = {'H': 51.63, 'He': -1.481,
```
PEP 8 does recommend that, but I think I would prefer not writing them in ALL_CAPS because they're basically data structures storing a lot of information rather than single constants. This is also in line with the dictionaries in `common.py`. I'm basically treating `data.py` as a file that contains no logic other than variable assignments, so similar to what a yaml or json file would be. I kept it as a Python file though because the data structure isn't particularly complicated and it allows for easy importing. Let me know if you still think I should change it or if we should consider storing the data in a different format altogether.

cgrambow commented on arkane/encorr/data.py:
```diff
@@ -0,0 +1,405 @@
+        'P': -341.2444299005 + SOC['P'], 'S': -398.0940312227 + SOC['S']
+    },
+
+    'klip_1': {
```
I was thinking to actually just completely replace unused model chemistries once we have the new BAC data. I highly doubt that anyone is using some of the more obscure ones and don't think anyone would miss them.

cgrambow commented on arkane/encorr/data.py:
```diff
@@ -0,0 +1,405 @@
+        'O': -74.995458316117 + SOC['O'], 'C': -37.788203485235 + SOC['C']
+    },
+
+    # We are assuming that SOC is included in the Bond Energy Corrections
```
We have two options: Either we add SOCs to all of the atom energies and then fit BACs or we don't add SOCs at all in which case they will be implicitly captured in the BACs. I prefer the second option.

cgrambow commented on arkane/encorr/data.py:
```diff
@@ -0,0 +1,405 @@
+        'I': -294.7958443 + SOC['I']
+    },
+
+    # note that all atom corrections but S are fitted, the correction for S is calculated
```
My plan was to have just one set of atom energies for both BAC methods and to just use the pure quantum calculations for the atoms at each level of theory as the atomic energies. In that way, all corrections due to systematic errors appear as part of the BACs. As far as I can tell, it's actually not very common in literature to fit both atom and bond corrections when using the conventional style of BACs, so I don't know why Arkane has been doing that thus far. For Melius-type BACs, corrections to the atomic energies are actually built in to the method.

cgrambow commented on arkane/encorr/data.py:
```diff
@@ -0,0 +1,405 @@
+        'O': -75.065397706471 + SOC['O'], 'C': -37.843634971592 + SOC['C']
+    },
+
+
```
How did you even catch this

cgrambow commented on arkane/encorr/mbac.py:
```diff
@@ -0,0 +1,141 @@
+def get_bac_correction(model_chemistry, coords, nums, multiplicity=1, mol_corr=0.0):
+    """
+    Given the model chemistry, molecular coordinates, atomic numbers,
+    and dictionaries of BAC parameters, return the total BAC correction
```
I was being inconsistent in other places as well, so I just went ahead and replaced all occurrences of "BAC correction" by "BAC", including in function names.

cgrambow commented:
I think I addressed all your comments and made some fixup commits. I just realized that I caused a circular import issue, so I'll fix that, too.

cgrambow commented:
Added two more fixup commits. I introduced an `exceptions.py` file in the main Arkane directory and added the exceptions there to get rid of the circular imports.

mliu49 commented on arkane/encorr/corr.py:
```diff
@@ -42,20 +42,31 @@
     it is consistent with the normal gas-phase reference states.
     Optionally, correct the energy using bond additivity corrections.
+
+    model_chemistry: The model chemistry, typically specified as method/basis.
```
You can format the arguments to get nice styling when we compile the API documentation. Minimum would be adding the `Args` keyword and indenting:
```
Args:
    model_chemistry: The model chemistry, typically specified as method/basis.
    ...
```

See https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html for even more details.

cgrambow replied:
Good to know!

alongd commented:
@cgrambow, can you rebase?
What else is needed for this PR?
Let's get it in before the Py3 transision

cgrambow commented:
Squashed and rebased. Nothing else is needed for this PR, except for BAC data, but we don't have that yet, so we can add it after Py3 transition.

alongd commented:
Could you take a look at the Codacy comment?

alongd commented on arkane/encorr/corr.py:
```diff
@@ -0,0 +1,118 @@
+        else:
+            raise Exception(
+                'Element "{}" is not yet supported in Arkane.'
+                ' To include it, add its experimental heat of formation'.format(symbol)
```
@cgrambow, was this comment addressed?

cgrambow replied:
Now it has been.

cgrambow commented:
The Codacy comment can be ignored. Apparently Codacy isn't smart enough to figure out that `c` in `coords` can be indexed.
